iree/hal/vulkan/dynamic_symbols.cc

```diff
@@ -92,6 +92,40 @@ Status ResolveFunctions(DynamicSymbols* syms,
   // single function.
   syms->vkGetInstanceProcAddr = reinterpret_cast<PFN_vkGetInstanceProcAddr>(
       get_proc_addr("vkGetInstanceProcAddr"));
+
+#if defined(IREE_PLATFORM_ANDROID)
+  // Since Android 8 Oreo, Android re-architected the OS frameowrk with project
+  // Treble. Framework libraries and vendor libraries have a more strict and
+  // clear separation. Their dependencies are carefully scrutinized and only
+  // selected cases are allowed. This is enforced with linker namespaces.
+  //
+  // /data/local/tmp is the preferred directory for automating native binary
+  // tests built using NDK toolchain. They should be allowed to access libraries
+  // like libvulkan.so for their functionality. However, there was an issue
+  // with fully treblized Android 10 where /data/local/tmp did not have access
+  // to the linker namespaces needed by libvulkan.so. This is fixed via
+  // https://android.googlesource.com/platform/system/linkerconfig/+/296da5b1eb88a3527ee76352c2d987f82f3252eb
+  //
+  // But as typically in the Android system, it takes a long time to see the
+  // fix getting propagated, if ever. A known workaround is to symlink the
+  // vendor Vulkan implementation under /vendor/lib[64]/hw/vulkan.*.so as
+  // libvulkan.so under /data/local/tmp and use LD_LIBRARY_PATH=/data/local/tmp
+  // when invoking the test binaries. This effectively bypasses the Android
+  // Vulkan loader. This is fine for ARM Mali GPUs, whose driver exposes
+  // the symbol `vkGetInstanceProcAddr`. But for Qualcomm Adreno GPUs,
+  // the Vulkan implementation library does not directly expose the symbol.
+  // Instead it's hidden as `qglinternal::vkGetInstanceProcAddr`. So try to
+  // see whether we can get this symbol. This is a reasonable workaround
+  // as otherwise it means we need to wrap. every. single. binary. test.
+  // as. a. full-blown. Android. app.
```

Comment:
the frustration is real -- I've been there :P
if we end up needing apk-associated junk (camera, microphone, etc) I have some internal code we can open up that does the wrapping and passes through command line arguments/etc - but the less we need to lean on that the better as things like `perf` won't work with apk-wrapped binaries :(

Reply:
👍 Yup, totally agreed. It's certainly not a fun thing to go through all those extra steps.. :)